rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/DocumentChanges.java

```diff
@@ -93,22 +88,41 @@ public static List<Either<TextDocumentEdit, ResourceOperation>> translateDocumen
                     // have to extend the entire/all LSP API with this information _per_ file?
                     result.add(Either.forLeft(
                         new TextDocumentEdit(new VersionedTextDocumentIdentifier(getFileURI(edit, "file"), null),
-                            translateTextEdits(docService, (IList) edit.get("edits")))));
+                            translateTextEdits(docService, (IList) edit.get("edits"), changeAnnotations))));
                     break;
             }
         }
 
-        return result;
+        WorkspaceEdit wsEdit = new WorkspaceEdit(result);
+        wsEdit.setChangeAnnotations(changeAnnotations);
+
+        return wsEdit;
     }
 
-    private static List<TextEdit> translateTextEdits(final IBaseTextDocumentService docService, IList edits) {
+    private static List<TextEdit> translateTextEdits(final IBaseTextDocumentService docService, IList edits, Map<String, ChangeAnnotation> changeAnnotations) {
         return edits.stream()
             .map(IConstructor.class::cast)
             .map(c -> {
+                var range = locationToRange(docService, (ISourceLocation) c.get("range"));
+                var replacement = ((IString) c.get("replacement")).getValue();
+                // Check annotation
                 var kw = c.asWithKeywordParameters();
-                return kw.hasParameter("annotation")
-                    ? new AnnotatedTextEdit(locationToRange(docService, (ISourceLocation) c.get("range")), ((IString) c.get("replacement")).getValue(), ((IString) kw.getParameter("annotation")).getValue())
-                    : new TextEdit(locationToRange(docService, (ISourceLocation) c.get("range")), ((IString) c.get("replacement")).getValue());
+                if (kw.hasParameter("annotation")) {
+                    var anno = (IConstructor) kw.getParameter("annotation");
+                    var label = ((IString) anno.get("label")).getValue();
+                    var description = ((IString) anno.get("description")).getValue();
+                    var needsConfirmation = ((IBool) anno.asWithKeywordParameters().getParameter("needsConfirmation")).getValue();
+                    var annoKey = String.format("%s_%s_%b", label, description, needsConfirmation);
```

Comment:
wouldn't `anno.hashCode()` be good enough? or do we need to send them somewhere?